pkg/containerutil/cp_linux.go

```diff
@@ -17,163 +17,200 @@
 	}
-	if container2host {
-		srcFull, err = securejoin.SecureJoin(root, src)
-		dstFull = dst
+
+	var sourceSpec, destinationSpec *pathSpecifier
+	var sourceErr, destErr error
+	if options.Container2Host {
+		sourceSpec, sourceErr = getPathSpecFromContainer(options.SrcPath, conSpec, root)
+		destinationSpec, destErr = getPathSpecFromHost(options.DestPath)
 	} else {
-		srcFull = src
-		dstFull, err = securejoin.SecureJoin(root, dst)
+		sourceSpec, sourceErr = getPathSpecFromHost(options.SrcPath)
+		destinationSpec, destErr = getPathSpecFromContainer(options.DestPath, conSpec, root)
 	}
-	if err != nil {
-		return err
+
+	if destErr != nil {
+		if errors.Is(destErr, errDoesNotExist) {
+			return ErrDestinationParentMustExist
+		} else if errors.Is(destErr, errIsNotADir) {
+			return ErrDestinationIsNotADir
+		}
+
+		return errors.Join(ErrFilesystem, destErr)
 	}
-	var (
-		srcIsDir       bool
-		dstExists      bool
-		dstExistsAsDir bool
-		st             fs.FileInfo
-	)
-	st, err = os.Stat(srcFull)
-	if err != nil {
-		return err
+
+	if sourceErr != nil {
+		if errors.Is(sourceErr, errDoesNotExist) {
+			return ErrSourceDoesNotExist
+		} else if errors.Is(sourceErr, errIsNotADir) {
+			return ErrSourceIsNotADir
+		}
+
+		return errors.Join(ErrFilesystem, sourceErr)
 	}
-	srcIsDir = st.IsDir()
 
-	// dst may not exist yet, so err is negligible
-	if st, err := os.Stat(dstFull); err == nil {
-		dstExists = true
-		dstExistsAsDir = st.IsDir()
+	// Now, resolve cp shenanigans
+	// First, cannot copy a non-existent resource
+	if !sourceSpec.exists {
+		return ErrSourceDoesNotExist
 	}
-	dstEndsWithSep := strings.HasSuffix(dst, string(os.PathSeparator))
-	srcEndsWithSlashDot := strings.HasSuffix(src, string(os.PathSeparator)+".")
-	if !srcIsDir && dstEndsWithSep && !dstExistsAsDir {
-		// The error is specified in https://docs.docker.com/engine/reference/commandline/cp/
-		// See the `DEST_PATH does not exist and ends with /` case.
-		return fmt.Errorf("the destination directory must exists: %w", err)
+
+	// Second, cannot copy into a readonly destination
+	if destinationSpec.readOnly {
+		return ErrTargetIsReadOnly
 	}
-	if !srcIsDir && srcEndsWithSlashDot {
-		return fmt.Errorf("the source is not a directory")
+
+	// Cannot copy a dir into a file
+	if sourceSpec.isADir && destinationSpec.exists && !destinationSpec.isADir {
+		return ErrCannotCopyDirToFile
 	}
-	if srcIsDir && dstExists && !dstExistsAsDir {
-		return fmt.Errorf("cannot copy a directory to a file")
+
+	// A file cannot be copied inside a non-existent directory with a trailing slash, or slash+dot
+	if !sourceSpec.isADir && !destinationSpec.exists && (destinationSpec.endsWithSeparator || destinationSpec.endsWithSeparatorDot) {
+		return ErrDestinationDirMustExist
 	}
-	if srcIsDir && !dstExists {
-		if err := os.MkdirAll(dstFull, 0o755); err != nil {
-			return err
+
+	// XXX FIXME: this seems wrong. What about ownership? We could be doing that inside a container
```

Comment:
I am not touching this right now.
This is previous code.
Am I mis-reading this, or is this a problem when doing that in a container target? (eg: shouldn't we do that as part of the nsenter/re-exec routine?)